### PR TITLE
video: avoid the build break in Visual Studio 2013 and earlier

### PR DESCRIPTION
**Pull request by tomoaki0705**
### This pullrequest changes
follow-up PR of #11092

### CODE REVIEW

File: modules/videoio/src/cap_msmf.cpp [hunk 1]
```diff
@@ -508,13 +508,15 @@ class Media_Foundation
...
+#if __cplusplus >= 201103L || (defined(_MSC_VER) && _MSC_VER >= 1900)
         //CXX11
         static thread_local Media_Foundation instance;
 #else //__cplusplus >= 201103L || _MSC_VER >= 1800
         //CXX98
-#ifdef _WIN32
+#if defined(_WIN32) && !defined(_MSC_VER)
         static __declspec(thread) Media_Foundation instance;
+#elif defined(_MSC_VER)
+        static Media_Foundation instance;
```
[the remark this answers is not in the document]
Comment (tomoaki0705) [reply modules/videoio/src/cap_msmf.cpp]: Thank you for the update.

Nice suggestion (I didn't know the ```TLS``` class), but I'm still facing a new problem.

Following your implementation, I get an error saying 
```
(715): error C2248: '`anonymous-namespace'::Media_Foundation::Media_Foundation' : cannot access private member declared in class '`anonymous-namespace'::Media_Foundation'
```
If I understand this correctly, in your code, ```TLS``` will try to access the constructor of ```Media_Foundation``` class, which is private.
So, only solution for this approach is to move the constructor to public [like this](https://github.com/tomoaki0705/opencv/blob/9d73827bdc4a96ca2a99dd22d78b41897f5802d7/modules/videoio/src/cap_msmf.cpp#L529-L532).

```
#if !(defined(_MSC_VER) && (_MSC_VER < 1900)) // TLS is only supported for POD in old MSVC
private:
#endif
    Media_Foundation(void) { CV_Assert(SUCCEEDED(MFStartup(MF_VERSION))); }
```

Am I following your opinion correctly ?  I hope we are on the same page.

[the remark this answers is not in the document]
Comment (alalek) [reply modules/videoio/src/cap_msmf.cpp]: 'public' is fine here. It is .cpp file, so this is not a public exposed API.
Or add 'friend class cv::TLSData<Media_Foundation>;'.

Also please remove all these "#if" - they are not needed with `cv::TLSData`.

**Comment by tomoaki0705** [comment]
Hmm.., the build machine is showing unnatural behavior.
The test says it failed, but each test says it passed....

On my local, the test fails explicitly, but not sure if this is related.

**Comment by alalek** [comment]
Looks like it is related to `MFShutdown()` calls.
@terfendail Could you please take a look on this?

**Comment by alalek** [comment]
Well, `cv::TLSData` releases underlying objects from other single "cleanup" thread. Looks like `MFShutdown()` should be called from the same thread as `MFStartup()`.

My suggestion is too remove this thread_local specifier at all.
[MSDN page](https://msdn.microsoft.com/en-us/library/windows/desktop/ms702238(v=vs.85).aspx) for `MFStartup()` doesn't require initialization for each thread.

```
static Media_Foundation& getInstance()
{
    static Media_Foundation instance;
    return instance;
}
```

(MSMF would work from the main thread only).


Alternatively, we can remove `MFShutdown()` call completely. Possible resource leaks during app shutdown are non-critical.

**Comment by tomoaki0705** [comment]
#11254 covers this PR.
I'll close this one

**Pull request closed without merging**